pkg/strategy/dca2/strategy.go

```diff
@@ -105,33 +105,77 @@ func (s *Strategy) Run(ctx context.Context, _ bbgo.OrderExecutor, session *bbgo.
 	s.Strategy.Initialize(ctx, s.Environment, session, s.Market, ID, s.InstanceID())
 	instanceID := s.InstanceID()
 
-	if s.Short {
-		s.openPositionSide = types.SideTypeSell
-		s.takeProfitSide = types.SideTypeBuy
-	} else {
-		s.openPositionSide = types.SideTypeBuy
-		s.takeProfitSide = types.SideTypeSell
-	}
-
 	if s.OrderGroupID == 0 {
 		s.OrderGroupID = util.FNV32(instanceID) % math.MaxInt32
 	}
 
 	// order executor
 	s.OrderExecutor.TradeCollector().OnPositionUpdate(func(position *types.Position) {
-		s.logger.Infof("position: %s", s.Position.String())
+		s.logger.Infof("[DCA] POSITION UPDATE: %s", s.Position.String())
 		bbgo.Sync(ctx, s)
 
 		// update take profit price here
+		s.updateTakeProfitPrice()
+	})
+
+	s.OrderExecutor.ActiveMakerOrders().OnFilled(func(o types.Order) {
+		s.logger.Infof("[DCA] FILLED ORDER: %s", o.String())
+		openPositionSide := types.SideTypeBuy
+		takeProfitSide := types.SideTypeSell
+		if s.Short {
+			openPositionSide = types.SideTypeSell
+			takeProfitSide = types.SideTypeBuy
+		}
+
+		switch o.Side {
+		case openPositionSide:
+			s.emitNextState(OpenPositionOrderFilled)
+		case takeProfitSide:
+			s.emitNextState(WaitToOpenPosition)
+		default:
+			s.logger.Infof("[DCA] unsupported side (%s) of order: %s", o.Side, o)
+		}
 	})
 
 	session.MarketDataStream.OnKLine(func(kline types.KLine) {
 		// check price here
+		if s.state != OpenPositionOrderFilled {
+			return
+		}
+
+		compRes := kline.Close.Compare(s.takeProfitPrice)
+		// price doesn't hit the take profit price
+		if (s.Short && compRes > 0) || (!s.Short && compRes < 0) {
+			return
+		}
+
+		s.emitNextState(OpenPositionOrdersCancelling)
 	})
 
 	session.UserDataStream.OnAuth(func() {
-		s.logger.Info("user data stream authenticated, start the process")
-		// decide state here
+		s.logger.Info("[DCA] user data stream authenticated")
+		time.AfterFunc(3*time.Second, func() {
```

Comment:
this might block the backtesting for 3 seconds?

Reply:
Yes.....